### DISCUSSION TITLE
Set tool table commands give error 20

### DISCUSSION THREAD

**Discussion by samitray** [grblHAL/core#530] [General]
Trying to set tool table values via G10 L1 P1 Z1.5 gives error 20. Any pointer where am i going wrong. I am on stm32F446 core and $341 is set to 0.

Also unable to get manual tool change working. After issuing T1M6 and on change of status to Tool, 0xA3 is sent, then cycle start signal is driven high. grblHAL state changes from tool to Idle, G01 line is then sent, immediately error 40 is returned. Since cycle start is done I assume once the state changes to Idle normal moves should work. No other commands are sent between 0xA3 and Cycle Strat

**Comment by terjeio** [comment]
> Trying to set tool table values via G10 L1 P1 Z1.5 gives error 20.

Tool table availability is a [compile time option](https://github.com/grblHAL/core/blob/0a86e57ddc622c758a2c490f4100af948beebee2/config.h#L510). Since you get error 20 I assume you have not enabled it.

> Also unable to get manual tool change working. 

Is this with ioSender?

**Reply by samitray** [reply]
> Tool table availability is a compile time option

Thanks.... had to be me overlooking simple config, report extension wiki mentions about the compile time option..

> Is this with ioSender?

Nope... A fork of Candle.... However when I directly hand feed all commands include Tool change Ack (0xA3) and Cycle start, grblHAL responds correctly. So it seems to be some timing or UI issue.
My logic is to watch the status message and check for transition of status into Tool, once that happens, Immediately stop sending further lines and send 0xA3... then watch status again to catch transition out of Tool (triggered by Cycle Start switch) at this point generally status is idle, the UI sends the next line and in response I get error 40, now it could be that there is no tool table so the M6 failed any ways. Let me setup tool table and test again.

**Reply by samitray** [reply]
@terjeio 
Some observations report extension wiki mentions

> [T:<tool number>,<axis offsets>]

I have N_Axis 6 and $# prints  [T:1|0.000,0.000,0.000,0.000,0.000,0.000|0.000]
what is the last value after second pipe. I tried G10 L1 P1 Z1.75  R2.5 
the R value did not show up..after the second pipe.. so I am guessing it is something else.
Does the R value get stored?

**Reply by samitray** [reply]
@terjeio 
Also when I tried to compile with N_TOOLS 32, the CubeIDE build fails. I get no response on serial, and a debug led blink inside driver_init also does not work. With N_TOOLS 16, everything is good.

**Reply by terjeio** [reply]
> I have N_Axis 6 and $# prints [T:1|0.000,0.000,0.000,0.000,0.000,0.000|0.000]
what is the last value after second pipe. 

Tip radius.

> Does the R value get stored?

No, not yet.

> Also when I tried to compile with N_TOOLS 32, the CubeIDE build fails.

An [assert](https://github.com/grblHAL/core/blob/0a86e57ddc622c758a2c490f4100af948beebee2/nvs_buffer.c#L218) fails due to insufficient [NVS storage](https://github.com/grblHAL/core/blob/0a86e57ddc622c758a2c490f4100af948beebee2/nvs.h#L30), you will have to increase it - e.g. 4096 is plenty. If you are using an EEPROM it has to be at least 32Kbit.
